### ISSUE TITLE
If the flyCam is disabled as part of the initialization in SimpleApplication, the cursor is forcefully set to visible during the appstates initialization

### ISSUE THREAD

**Issue by riccardobl** [jMonkeyEngine/jmonkeyengine#1219]
The following code doesnt work as expected when called from simpleInitApp, since the FlyBy cam appstate will force inputManager.setCursorVisible(**true**) during the initialization of the appstates
```java
flyCam.setEnabled(false);
inputManager.setCursorVisible(false);
```
A temporary workaround is to also detach the appstate
```java
flyCam.setEnabled(false);
inputManager.setCursorVisible(false);
stateManager.detach(stateManager.getState(FlyCamAppState.class));
```

**Comment by qs-dragon** [comment]
This appears to be caused when the flycam appstate creates the new flybycamera, the flybycamera defaults it's internal "enabled" to true. Would adding a constructor that accepts an argument to set enabled, and change flycamappstate to use this constructor be a good approach?

**Comment by stephengold** [comment]
This is an annoying bug, and unfortunately Riccardo's workaround doesn't work for my application.

Adding an argument to the `FlyCamAppState` constructor seems like a good approach.

**Comment by pspeed42** [comment]
The work around doesn't work for you because you partially want the fly cam?

Note: if applications use the preferred method of specifying app states on the super constructor (and then not including FlyCamAppState) they don't even have to disable flyCam because it will never even be created.

Anyway, I agree with a constructor argument. It's a simple thing to partially fix these horrible classes.

**Comment by qs-dragon** [comment]
After trying some things, I think the issue is abit deeper than I originally thought. I cannot remove this behaviour without breaking changes to the flycam or the flycam appstate...

**Comment by pspeed42** [comment]
I can't help if you don't explain.

**Comment by pspeed42** [comment]
I think the issue with the code as you've committed it is that it's backwards.  The app state should be dictating the enabled state of the camera not the other way around.  And it shouldn't be an else branch.

If fly cam not created, create one.
Set the fly cam state to the app state's enabled state...not on an else branch.  Just always do it.

And then don't change anything about FlyCam.java.

Then callers can temporarily disable the flyCam by disabling the app state.

**Comment by qs-dragon** [comment]
I agree the flycam behaviour should be controlled from the appstate and not from flycam, but SimpleApplication only exposes flyCam and not the state. This makes it difficult to deal with since simpleInit is called before the flyCamAppState gets initialized, hence I tried to do it "backwards" to see if it worked, but the problem is now also "backwards": cursor is visible by default when it should start invisible, but toggled correctly otherwise.

**Comment by pspeed42** [comment]
Looking through the code again, I remember the compromises that I had to make to put this in an app state where it should have been in the first place.  I remember now why only the super() constructor can be used to not register the app state. (Actually, looking again, removing the app state in simpleInit() should still be a way to skip the cursor stuff.)

The thing is in the case of SimpleApplication, the block in the FlyCamAppState that constructs the flyCam is never used... because SimpleApplication creates its own flyCam.  (which is also because FlyByCamera won't let you switch cameras... so it must be provided on the constructor instead of later... dumb.)

The changes needed here to really fix things are deeper.  But maybe there is a short-cut to just fix this issue.

First, I think we should decide that directly setting flyCam.setEnabled(true/false) should not work for SimpleApplication's flyCam.  Only stateManager.getState(FlyCamAppState.class).setEnabled(false) should be used to disable SimpleApplication's flyCam.

Then we need to fix FlyCamAppState because it's kind of broken.  It sets the cursor on initialization.

FlyCamAppState should be changed to extend BaseAppState.  Most of the code gets removed as initialize() and cleanup won't be doing anything and it won't be managing its own setEnabled()/isEnabled() anymore.  It should override onEnable() and onDisable() to set the fly cam enabled/disabled and to register/unregister for input.  (FlyByCamera is such a mess that you can disable it but it will still gobble input... but I don't think we can change that now.)

Given that the app state will always be enabling/registering input or disabling/unregistering input... it's probably ok to leave the crappy code that does the setCursorVisible() in two different places.  We're hindered by the fact that FlyByCamera doesn't have an app state's lifecycle (should have been an app state all along).

Hmmm... I wonder how many problems it would solve to have FlyByCam extend BaseAppState and then properly implement onEnable()/onDisable().  I guess technically, to not break user code, FlyCamAppState would have to be in the mix also.  But it would solve a lot of problems if FlyByCamera had an app state's lifecycle.  

The only down side to me is that something from jme3.input would extend something from jme3.app... but to me, FlyByCamera should have been an app state all along and is in the wrong place anyway.  While we can't fix that, the cross package dependency seems the least ugly solution.

Off the top of my head, FlyByCamera extends BaseAppState, FlyCamAppState extends FlyByCamera and its getCamera() method just returns "this".

FlyByCamera then moves its setEnabled() stuff into onEnable()/onDisable() and just lets base app state manage the enabled flag.  onEnable()/onDisable() can also registerWithInput()/unregisterInput().  Also add a no-arg constructor so that the camera can be set later.

Then in SimpleApplication, instead of creating the flyCam we'd just set it to the FlyCamAppState and set its Camera.
